File: content/problems/2267-check-if-there-is-a-valid-parentheses-string-path/solutions/grid-balance-state-dp/solution.cpp
class Solution {
    public:bool hasValidPath(vector<vector<char>>& grid){
        int rows=grid.size(),columns=grid[0].size(),length=rows+columns-1;
        if(length%2||grid[0][0]!='('||grid[rows-1][columns-1]!=')')return false;
        vector reachable(rows,vector(columns,vector<bool>(length+2)));
        reachable[0][0][1]=true;
        int directions[3]={
            1,0,1
        };
        for(int r=0;r<rows;++r)for(int c=0;c<columns;++c)for(int balance=0;balance<=length;++balance)if(reachable[r][c][balance])for(int d=0;d<2;++d){
            int nr=r+directions[d],nc=c+directions[d+1];
            if(nr>=rows||nc>=columns)continue;
            int next=balance+(grid[nr][nc]=='('?1:-1);
            if(next>=0&&next<=length)reachable[nr][nc][next]=true;
        }
        return reachable[rows-1][columns-1][0];
    }
};

File: content/problems/2267-check-if-there-is-a-valid-parentheses-string-path/solutions/grid-balance-state-dp/Solution.kt
class Solution {
    fun hasValidPath(grid:Array<CharArray>):Boolean{
        val rows=grid.size;
        val columns=grid[0].size;
        val length=rows+columns-1;
        if(length%2==1||grid[0][0]!='('||grid[rows-1][columns-1]!=')')return false;
        val reachable=Array(rows){
            Array(columns){
                BooleanArray(length+2)
            }
        };
        reachable[0][0][1]=true;
        val directions=arrayOf(intArrayOf(1,0),intArrayOf(0,1));
        for(r in 0 until rows)for(c in 0 until columns)for(balance in 0..length)if(reachable[r][c][balance])for(direction in directions){
            val nr=r+direction[0];
            val nc=c+direction[1];
            if(nr>=rows||nc>=columns)continue;
            val next=balance+if(grid[nr][nc]=='(')1 else -1;
            if(next in 0..length)reachable[nr][nc][next]=true
        };
        return reachable[rows-1][columns-1][0]
    }
}
